### PR TITLE
IPC Organ Printing

### PR DESCRIPTION
**Pull request by CASEYGUNGEON**
Version 1: It Works on my End!

## About The Pull Request

Apparently Exosuit Fabricators could build IPC organs before, but now they can't. Except NOW now they can. Cheap early-game tech.

## How This Contributes To The Skyrat Roleplay Experience

IPCs can get missing organs put back in them, organics can get their organs replaced, (tested; does in fact let organics interface with APCs with only organ replacement) and maybe IPCs can be made from scratch.

## Changelog

:cl:
add: Early-game tech allows Exosuit Fabs to print IPC organs.
/:cl:

### CODE REVIEW

**Comment by CASEYGUNGEON** [comment]
Oh, whoops. I didn't know updating my fork would mess with this pull request. I guess I need to reverse that commit or something.

**Comment by CASEYGUNGEON** [comment]
Sorry for the mess. First time.

**Comment by CASEYGUNGEON** [comment]
Relatively minor thing, but i noticed the pre-req isn't actually the one i wrote down in the PR. it's the one that starts already done that's called like Basic Medical Equipment. I don't really think that's an issue so I didn't push anything else.

File: code/modules/vehicles/mecha/mech_fabricator.dm [hunk 1]
```diff
@@ -191,6 +194,13 @@
 		else if(ispath(built_item, /obj/item/borg_restart_board))
 			sub_category += "All Cyborgs" //Otherwise the restart board shows in the "parts" category, which seems dumb
 
+		// SKYRAT EDIT START - ipc_printing
+
+		else if(findtext(built_item, "robot_ipc") || findtext(built_item,"power_cord"))
```
Comment (GoldenAlpharex) [inline code/modules/vehicles/mecha/mech_fabricator.dm:199]: I got a feeling that's probably not the best way to do this, have you tried using `ispath()` like the other ones?

File: modular_skyrat/modules/ipc_printing/code/ipc_designs.dm [hunk 2]
```diff
@@ -0,0 +1,55 @@
+/datum/design/ipc_parts
+	name = "IPC Part"
+	build_type = MECHFAB
+	departmental_flags = DEPARTMENTAL_FLAG_SCIENCE
+	materials = list(/datum/material/iron=1000)
```
Comment (GoldenAlpharex) [inline modular_skyrat/modules/ipc_printing/code/ipc_designs.dm]:
```suggestion
	materials = list(/datum/material/iron = 1000)
```

File: modular_skyrat/modules/ipc_printing/code/ipc_designs.dm [hunk 3]
```diff
@@ -0,0 +1,55 @@
...
+	construction_time = 50
```
Comment (GoldenAlpharex) [inline modular_skyrat/modules/ipc_printing/code/ipc_designs.dm]:
```suggestion
	construction_time = 5 SECONDS
```

File: modular_skyrat/modules/ipc_printing/code/ipc_printing_techweb.dm [hunk 4]
```diff
@@ -0,0 +1,16 @@
+/datum/techweb_node/ipc_parts
+	id = "ipc_parts"
+	display_name = "IPC Organs"
+	description = "Allows Exosuit Fabricators to build IPC organs."
+	prereq_ids = list("basic_medical")
```
Comment (GoldenAlpharex) [inline modular_skyrat/modules/ipc_printing/code/ipc_printing_techweb.dm]: That means it'll be available to be researched by default. I'd probably move that to be after something as part of the robotics research tree.

**Review by GoldenAlpharex — CHANGES_REQUESTED** [review]
Good change, only a few things to look through.

File: code/modules/vehicles/mecha/mech_fabricator.dm (hunk 1 above)
Comment (CASEYGUNGEON) [reply code/modules/vehicles/mecha/mech_fabricator.dm:199]: iirc, using `ispath()` also hits the cyber organs and changes their category.

File: modular_skyrat/modules/ipc_printing/code/ipc_printing_techweb.dm (hunk 4 above)
Comment (CASEYGUNGEON) [reply modular_skyrat/modules/ipc_printing/code/ipc_printing_techweb.dm]: yeah, makes sense. i'd originally meant it to have the same pre-req as cyber organs but read the labels wrong at the time. i'll take another look at them.

Comment (CASEYGUNGEON) [reply modular_skyrat/modules/ipc_printing/code/ipc_printing_techweb.dm]: changed the pre-req to `adv_robotics`, considering the description.

**Comment by CASEYGUNGEON** [comment]
not sure why it's failing trying to make/destroy shadow bodyparts. i guess i'll try and take a look at that later today.

**Comment by Zergspower** [comment]
Does this add their sprites in? Currently half of them dont have them

**Comment by CASEYGUNGEON** [comment]
Just checked. Only the reagent processing unit doesn't have a sprite. guess i just never counted when i tried to print all of them.

i gotta draw up a sprite or something.

**Comment by CASEYGUNGEON** [comment]
Good news! the reason why the liver didn't have a sprite is because it was looking for `liver-c` instead of `liver-ipc`. i guess technically that's a change so i added it to the readme and put a comment next to where I fixed it. All 8 organs have sprites.

**Comment by CASEYGUNGEON** [comment]
reorganized my fork, didn't know it would auto-close the PR. can't reopen this, but i can open a second PR if that's not too weird. shouldn't have to reorganize again later. (this PR was pulling from my main branch. wanted to be able to fetch upstream and try other changes and not have this stuff be in my main)

**Pull request closed without merging**